opm/ml/keras_model.cpp

```diff
@@ -0,0 +1,419 @@
+    KASSERT(ReadFloat(file, &data_max), "Failed to read max");
+    KASSERT(ReadFloat(file, &feat_inf), "Failed to read max");
+    KASSERT(ReadFloat(file, &feat_sup), "Failed to read max");
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerScaling<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    KASSERT(in, "Invalid input");
+    KASSERT(out, "Invalid output");
+
+    Tensor<Evaluation> temp_in, temp_out;
+
+    *out = *in;
+
+    for (size_t i = 0; i < out->data_.size(); i++) {
+        auto tempscale = (out->data_[i] - data_min)/(data_max - data_min);
+        out->data_[i] = tempscale * (feat_sup - feat_inf) + feat_inf;
+    }
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerUnScaling<Evaluation>::LoadLayer(std::ifstream* file) {
+    KASSERT(file, "Invalid file stream");
+
+    KASSERT(ReadFloat(file, &data_min), "Failed to read min");
+    KASSERT(ReadFloat(file, &data_max), "Failed to read max");
+    KASSERT(ReadFloat(file, &feat_inf), "Failed to read inf");
+    KASSERT(ReadFloat(file, &feat_sup), "Failed to read sup");
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerUnScaling<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    KASSERT(in, "Invalid input");
+    KASSERT(out, "Invalid output");
+
+    *out = *in;
+
+    for (size_t i = 0; i < out->data_.size(); i++) {
+        auto tempscale = (out->data_[i] - feat_inf)/(feat_sup - feat_inf);
+
+        out->data_[i] = tempscale * (data_max - data_min) + data_min;
+    }
+
+
+    return true;
+}
+
+
+template<class Evaluation>
+bool KerasLayerDense<Evaluation>::LoadLayer(std::ifstream* file) {
+    KASSERT(file, "Invalid file stream");
+
+    unsigned int weights_rows = 0;
+    KASSERT(ReadUnsignedInt(file, &weights_rows), "Expected weight rows");
+    KASSERT(weights_rows > 0, "Invalid weights # rows");
+
+    unsigned int weights_cols = 0;
+    KASSERT(ReadUnsignedInt(file, &weights_cols), "Expected weight cols");
+    KASSERT(weights_cols > 0, "Invalid weights shape");
+
+    unsigned int biases_shape = 0;
+    KASSERT(ReadUnsignedInt(file, &biases_shape), "Expected biases shape");
+    KASSERT(biases_shape > 0, "Invalid biases shape");
+
+    weights_.Resize(weights_rows, weights_cols);
+    KASSERT(
+        ReadFloats(file, weights_.data_.data(), weights_rows * weights_cols),
+        "Expected weights");
+
+    biases_.Resize(biases_shape);
+    KASSERT(ReadFloats(file, biases_.data_.data(), biases_shape),
+            "Expected biases");
+
+    KASSERT(activation_.LoadLayer(file), "Failed to load activation");
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerDense<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    KASSERT(in, "Invalid input");
+    KASSERT(out, "Invalid output");
+    KASSERT(in->dims_.size() <= 2, "Invalid input dimensions");
+
+    if (in->dims_.size() == 2) {
```

Comment:
The dimension things here require an explanation. Why can we not apply a dense layer to a 1-tensor, and why is there a special treatment for 2-tensors here?